### ISSUE TITLE
Allow custom Dockerfile

### ISSUE THREAD

**Issue by clabu** [metacloud/molecule#393]
Currently the Dockerfile used to customize the image is hard-coded in `molecule/driver/dockerdriver.py`. Allow the option for a custom Dockerfile. I for example need the image to have a corporate proxy configured or I cannot download packages. I would also like an image with init system working. 

Example config:
```
---
docker:
  containers:
    - name: foo-01
      ansible_groups:
      - group1
      dockerfile: tests/Dockerfile.centos-6
      privileged: True
      port_bindings:
        80: 80
    - name: foo-02
      ansible_groups:
        - group2
      image: ubuntu
      image_version: '14.04'
      registry: testhost:5323
      volume_mounts:
        - '/this/volume:/to/this:rw'
      command: '/bin/sh'
```

**Comment by kireledan** [comment]
This seems a bit unnecessary. If you build the image locally with your customized dockerfile you can already set that as your specified image. Also I'm a bit confused by your config. You're passing a dockerfile but set the image to something else. 

Do you mean to say you would want to only pass in the dockerfile with no image?

**Comment by clabu** [comment]
Yes, I mean that only pass the dockerfile setting and no image setting. Sorry for the poor example; I will update it.

My use case is that I want to check in the Dockerfile with my tests so that folks that do not have access to my Docker registry can still build the custom image. I guess the alternative is to execute the same commands in playbook.yml, but they would then be done on each test run (slow).

On the other hand, this use case does not translate to OpenStack or Vagrant where any "image" customization would still have to be done in playbook.yml. But still, I think would be convenient for some Docker users.

**Comment by retr0h** [comment]
I'm not entirely sure I'm sold on this.  Not sure we want to get into the game of building docker images.  We currently "modify" pre-built images to ensure they work in molecule, but building an image off a custom docker file would be a bit out of scope for molecule.

Instead, have you thought of using tools to build the image?  This would be fairly trivial.  One example is [ansible-container](https://github.com/ansible/ansible-container)?  This will allow you to code'ify your docker files, and have them uploaded into the developers local registry,  where molecule could then utilize it.

**Comment by kireledan** [comment]
I think I can definitely see a use case for this though. I was able to whip something up quickly if you want to take a look at it?

**Comment by clabu** [comment]
Building an image for my Molecule tests using ansible-container would be fine. However, it'd be great if this could be done in the normal Molecule workflow so that users can just get the role and run "molecule test" without first having to build a Docker image with a separate command. 

How about a pre-hook to run a playbook before Molecule starts creating instances? In my case the pre-hook playbook would run ansible-container to build the image that Molecule will use in the "Creating instances" next step.

```
$ molecule test
--> Destroying instances ...
--> Checking playbooks syntax ...

playbook: playbook.yml
--> Running pre-hook playbook ... # NEW
--> Creating instances ...
```

**Comment by retr0h** [comment]
> I think I can definitely see a use case for this though. I was able to whip something up quickly if you want to take a look at it?

Sure

**Comment by retr0h** [comment]
This was implemented with #409.  Will be available on pypi when we release 1.10.

**Comment by clabu** [comment]
Awesome!
